rlmolecule/mcts/mcts.py

```diff
@@ -0,0 +1,173 @@
+
+        def ucb_selection(parent):
+            return max(parent.children, key=lambda child: self._ucb_score(parent, child))
+
+        mcts_selection_function = ucb_selection if mcts_selection_function is None else mcts_selection_function
+
+        # noinspection PyTypeChecker
+        yield from self.run_from_vertex(vertex, action_selection_function, num_mcts_samples, mcts_selection_function)
+
+    def run_from_vertex(
+            self,
+            vertex: MCTSVertex,
+            action_selection_function: Callable[[MCTSVertex], MCTSVertex],
+            num_mcts_samples: int,
+            mcts_selection_function: Callable[[MCTSVertex], MCTSVertex],
+    ) -> Generator[MCTSVertex]:
+        yield vertex
+        self.sample(vertex, num_mcts_samples, mcts_selection_function)
+        # self.problem.store_policy_inputs_and_targets(state) # elided
+        children = vertex.children
+        if children is not None and len(children) > 0:
+            child = action_selection_function(vertex)
+            yield from self.run_from_vertex(child, action_selection_function, num_mcts_samples, mcts_selection_function)
+
+    def sample(
+            self,
+            vertex: MCTSVertex,
+            num_mcts_samples: int,
+            mcts_selection_function: Callable[[MCTSVertex], MCTSVertex],
+    ) -> None:
+        for _ in range(num_mcts_samples):
+            search_path, value = self._select(vertex, mcts_selection_function)
+            self._backpropagate(search_path, value)
+
+    def _select(
+            self,
+            root: MCTSVertex,
+            mcts_selection_function: Callable[[MCTSVertex], MCTSVertex],
+    ) -> ([MCTSVertex], float):
+        """
+        Selection step of MCTS
+        From Wikipedia (https://en.wikipedia.org/wiki/Monte_Carlo_tree_search):
+        Selection: Start from root R and select successive child vertices until a leaf vertex L is reached.
+        The root is the current game state and a leaf is any vertex that has a potential child from which no simulation
+        (playout) has yet been initiated. The section below says more about a way of biasing choice of child vertices that
+        lets the game tree expand towards the most promising moves, which is the essence of Monte Carlo tree search.
+        """
+        current = root
+        search_path = []
+        while True:
+            search_path.append(current)
+            children = current.children
+
+            if children is None:  # node is unexpanded: expand and return its value estimate
+                current.children = [self.get_vertex_for_state(state) for state in current.state.get_next_actions()]
+                return search_path, self._evaluate(current, search_path)
+
+            if len(children) == 0:  # node is expanded and terminal: return its value
+                return search_path, self.problem.get_reward(current.state)
+
+            current = mcts_selection_function(current)
```

Comment:
this was previously called `tree_policy` I'm assuming? I had it set up as a recursive function, but there's no reason we have to keep it that way if the while loop gives us more flexibility. Something with odd MCTS though is that you only expand if a node has been previously visited -- this looks like it would expand a child node any time it's visited?